src/common/iop_order.c

```diff
@@ -1411,6 +1413,144 @@ static void _ioppr_check_rules(GList *iop_list, const int imgid, const char *msg
+  DT_DEBUG_SQLITE3_BIND_INT(stmt, 1, imgid);
+  if(sqlite3_step(stmt) == SQLITE_ROW) history_size = sqlite3_column_int(stmt, 0);
+  sqlite3_finalize(stmt);
+
+  if (history_size <1)
+  {
+    fprintf(stderr,"\n[dt_ioppr_convert_onthefly] for image %i has no valid history\n",imgid);
+    return;
+  }
+
+  GList *current_iop_list = dt_ioppr_get_iop_order_list(NULL);
+
+  // get the number of known iops
+  int valid_iops = 0;
+
+  GList *iops_order = g_list_last(current_iop_list);
+  while(iops_order)
+  {
+    valid_iops++; 
+    iops_order = g_list_previous(iops_order);
+  }
+
+  if (DT_ONTHEFLY_INFO) fprintf(stderr,"\n*** checking for known iops ***\n");
+
+  valid_iops = 0; // reuse it as an index  
+  iops_order = g_list_last(current_iop_list);
+  while(iops_order)
+  {
+    dt_iop_order_entry_t *order_entry = (dt_iop_order_entry_t *)iops_order->data;
+    if (DT_ONTHEFLY_INFO) fprintf(stderr,"  %s, %f\n",order_entry->operation,order_entry->iop_order);
+    valid_iops++; 
+    iops_order = g_list_previous(iops_order);
+  }
+
+  typedef struct dt_onthefly_history_t
+  {
+    int num;
+    char operation[20];
+    double old_iop_order;
+    double new_iop_order;
+    int multi_priority;
+  } dt_onthefly_history_t;
+  struct dt_onthefly_history_t *myhistory = (dt_onthefly_history_t *)calloc(history_size, sizeof(dt_onthefly_history_t));
+
+  // read in the history
+  for (int i=0;i<history_size;i++)
+  {
+    int hits=0;
+    DT_DEBUG_SQLITE3_PREPARE_V2(dt_database_get(darktable.db),
+      "SELECT num, operation, iop_order, multi_priority FROM main.history WHERE imgid=?1", -1, &stmt, NULL);
+    DT_DEBUG_SQLITE3_BIND_INT(stmt, 1, imgid);
+    while(sqlite3_step(stmt) == SQLITE_ROW)
+    { 
+      struct dt_onthefly_history_t *this = &myhistory[hits];
+      this->num = sqlite3_column_int(stmt, 0);
+      g_strlcpy(this->operation, (char *)sqlite3_column_text(stmt, 1), 20);
+      this->old_iop_order = this->new_iop_order = sqlite3_column_double(stmt,2);
+      this->multi_priority = sqlite3_column_int(stmt, 3);      
+      hits++;
+    }
+    sqlite3_finalize(stmt);
+  }
+
+  // process history
+  for (int i=0;i<history_size;i++)
+  {
+    struct dt_onthefly_history_t *this = &myhistory[i];
+    this->new_iop_order = dt_ioppr_get_iop_order(current_iop_list, this->operation) + (double)this->multi_priority / 100.0;
+  }
+
+  // print complete history information 
+  fprintf(stderr,"\n\n ***** On-the-fly history V[%i]->V[%i], imageid: %i ****************",my_iop_order_version,DT_IOP_ORDER_VERSION,imgid);  
+  for (int i=0;i<history_size;i++)
+  {
+    struct dt_onthefly_history_t *this = &myhistory[i];
+    fprintf(stderr,"\n %3i %20s multi%3i :: iop %14.11f -> %14.11f",this->num,this->operation,this->multi_priority,this->old_iop_order,this->new_iop_order);
+  }
+
+  // really write history?, not yet
+  if (DT_ONTHEFLY_WRITING)
+  {
+    for (int i=0;i<history_size;i++)
```

Comment:
I would wrap this loop inside:

```
sqlite3_exec(db->handle, "BEGIN TRANSACTION", NULL, NULL, NULL);
if (...)
{
}
sqlite3_exec(db->handle, "COMMIT", NULL, NULL, NULL);
```

To ensure that the whole image is updated or nothing is done in case of crash.

Reply:
How about the main.images /iop_order_version change? Can/should that also be included before sqlite3_exec?

Reply:
main.images update should be part of the same transaction.

Reply:
done